kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/records/RecordBatchUtils.java

```diff
@@ -85,41 +81,40 @@ private RecordBatchUtils() {
      * applying a per-record mapping operation to each record.
      * @param recordBatch The batch of records to convert
      * @param mapper The mapping function to apply to the records in the batch
-     * @param resultBuffer A final buffer. This is required to encourage buffer reuse when
+     * @param builder The final builder. This is required to encourage buffer reuse when
      * callers make multiple invocations of this method.
      * @return A MemoryRecords containing the mapped records
      */
-    public static @NonNull MemoryRecords toMemoryRecords(
-                                                         @NonNull RecordBatch recordBatch,
-                                                         @NonNull RecordTransform mapper,
-                                                         @NonNull ByteBufferOutputStream resultBuffer) {
+    public static @NonNull BatchAwareMemoryRecordsBuilder toMemoryRecords(
+                                                                          @NonNull RecordBatch recordBatch,
+                                                                          @NonNull RecordTransform mapper,
+                                                                          @NonNull BatchAwareMemoryRecordsBuilder builder) {
```

Comment:
I was a bit sad that these changes were necessary, but it's a consequence of how the InBandKeyManager is now decompressing the records a batch at a time: We want there to be a single BAMRB, but this method meant we had a BAMRB for each batch.

This is a bit of a dog's breakfast right now, sharing a BAMRB across multiple calls to this method doesn't really fit the collector pattern. 

Perhaps we should drop the Stream idea, but keep a utility method for encapsulating it, rather than trying to conform to `Collector`, That would allow us to make the record's `index` a first class part of the `RecordTransformer` interface, as I alluded to [here](https://github.com/kroxylicious/kroxylicious/pull/899#discussion_r1465840824).

Reply:
> Perhaps we should drop the Stream idea, but keep a utility method for encapsulating it, rather than trying to conform to Collector, That would allow us to make the record's index a first class part of the RecordTransformer interface, as I alluded to https://github.com/kroxylicious/kroxylicious/pull/899#discussion_r1465840824.

Should the index be first class? Maybe the first class bit should be the state that's getting associated with the record? I could imagine arranging it so the user supplies a `Function<Record, State>` to map all their records, then we encapsulate the code that tracks `batchIndex -> state` so that when transforming the util code can call `transformer.init(record, state)`, rather than having each transformer implementation knowing how to retrieve state by index.

Reply:
This looks like a great idea, I think we should try it.